librz/analysis/arch/x86/x86_il.c

```diff
@@ -0,0 +1,3556 @@
+	"bp", /* X86_REG_BP */
+	"si", /* X86_REG_SI */
+	"di", /* X86_REG_DI */
+	NULL
+};
+
+/**
+ * \brief All registers bound to IL variables for x86 32-bit
+ */
+static const char *x86_bound_regs_32[] = {
+	COMMON_REGS,
+	"eax", /* X86_REG_EAX */
+	"ebx", /* X86_REG_EBX */
+	"ecx", /* X86_REG_ECX */
+	"edx", /* X86_REG_EDX */
+	"eip", /* X86_REG_EIP */
+	"esp", /* X86_REG_ESP */
+	"ebp", /* X86_REG_EBP */
+	"esi", /* X86_REG_ESI */
+	"edi", /* X86_REG_EDI */
+	"nt", /* X86_EFLAGS_NT */
+	"rf", /* X86_EFLAGS_RF */
+	"vm", /* X86_EFLAGS_VM */
+	"ac", /* X86_EFLAGS_AC */
+	"fs", /* X86_REG_FS */
+	"gs", /* X86_REG_GS */
+	"cr0", /* X86_REG_CR0 */
+	"dr0", /* X86_REG_DR0 */
+	NULL
+};
+
+/**
+ * \brief All registers bound to IL variables for x86 64-bit
+ */
+static const char *x86_bound_regs_64[] = {
+	COMMON_REGS,
+	"rax", /* X86_REG_RAX */
+	"rbx", /* X86_REG_RBX */
+	"rcx", /* X86_REG_RCX */
+	"rdx", /* X86_REG_RDX */
+	"rip", /* X86_REG_RIP */
+	"rsp", /* X86_REG_RSP */
+	"rbp", /* X86_REG_RBP */
+	"rsi", /* X86_REG_RSI */
+	"rdi", /* X86_REG_RDI */
+	"nt", /* X86_EFLAGS_NT */
+	"rf", /* X86_EFLAGS_RF */
+	"vm", /* X86_EFLAGS_VM */
+	"ac", /* X86_EFLAGS_AC */
+	"fs", /* X86_REG_FS */
+	"gs", /* X86_REG_GS */
+	"cr0", /* X86_REG_CR0 */
+	"dr0", /* X86_REG_DR0 */
+	NULL
+};
+
+static const X86Reg gpr_hregs[] = {
+	X86_REG_AH, // rax
+	X86_REG_BH, // rbx
+	X86_REG_CH, // rcx
+	X86_REG_DH, // rdx
+	X86_REG_INVALID, // rbp
+	X86_REG_INVALID, // rdi
+	X86_REG_INVALID, // rip
+	X86_REG_INVALID, // riz
+	X86_REG_INVALID, // rsi
+	X86_REG_INVALID // rsp
+};
+
+static const X86Reg gpr_lregs[] = {
+	X86_REG_AL, // rax
+	X86_REG_BL, // rbx
+	X86_REG_CL, // rcx
+	X86_REG_DL, // rdx
+	X86_REG_BPL, // rbp
+	X86_REG_DIL, // rdi
+	X86_REG_INVALID, // rip
+	X86_REG_INVALID, // riz
+	X86_REG_SIL, // rsi
+	X86_REG_SPL, // rsp
+};
+
+static const X86Reg gpr_xregs[] = {
+	X86_REG_AX, // rax
+	X86_REG_BX, // rbx
+	X86_REG_CX, // rcx
+	X86_REG_DX, // rdx
+	X86_REG_BP, // rbp
+	X86_REG_DI, // rdi
+	X86_REG_IP, // rip
+	X86_REG_INVALID, // riz
+	X86_REG_SI, // rsi
+	X86_REG_SP, // rsp
+};
+
+static const X86Reg gpr_eregs[] = {
+	X86_REG_EAX, // rax
+	X86_REG_EBX, // rbx
+	X86_REG_ECX, // rcx
+	X86_REG_EDX, // rdx
+	X86_REG_EBP, // rbp
+	X86_REG_EDI, // rdi
+	X86_REG_EIP, // rip
+	X86_REG_EIZ, // riz
+	X86_REG_ESI, // rsi
+	X86_REG_ESP, // rsp
+};
+
+static const X86Reg gpr_rregs[] = {
+	X86_REG_RAX,
+	X86_REG_RBX,
+	X86_REG_RCX,
+	X86_REG_RDX,
+	X86_REG_RBP,
+	X86_REG_RDI,
+	X86_REG_RIP,
+	X86_REG_RIZ,
+	X86_REG_RSI,
+	X86_REG_RSP
+};
+
+static bool x86_il_is_gpr(X86Reg reg) {
+	for (unsigned int i = 0; i < GPR_FAMILY_COUNT; i++) {
+		if (reg == gpr_hregs[i] || reg == gpr_lregs[i] || reg == gpr_xregs[i] || reg == gpr_eregs[i] || reg == gpr_rregs[i]) {
+			return true;
+		}
+	}
+
+	return false;
+}
+
+static ut8 x86_il_get_reg_size(X86Reg reg) {
+	for (unsigned int i = 0; i < GPR_FAMILY_COUNT; i++) {
+		if (reg == gpr_hregs[i] || reg == gpr_lregs[i]) {
+			return 8;
```

Comment:
Please add an asm test to cover this branch if possible.

Reply:
This is not practically possible, since for this I need to have an instruction which reads memory at `AL` or `AH` (8-bit registers), but it is invalid to use them, since they would point to invalid memory locations. So basically it is not possible to have any valid instruction which covers this path.

Reply:
Makes sense, such specific cases can indeed be left as-is.